convert tsfeatures lib to nbdev framework

jope35 opened the pull request:
This pr uses the nbdev framework to build the tsfeatures lib. 
All old files are copied 1-on-1, eager to get some feedback on this

jope35 committed:
convert lib to nbdev

jmoralez commented:
Hey @jope35, thanks for the initiative, we support this. Just some requests before we review this more thoroughly:

* Please restore the files under the `.github` folder.
* Please restore the .gitignore, LICENSE and setup.py files.
* Please remove the `pre-commit-config.yaml` and `.ruff.toml`. These can be added in a later PR so that we can focus on those changes specifically.
* Please restore and move the `compare_with_r.py` file to a `scripts` folder or similar.

jope35 committed:
remove pre-commit and .ruff

jope35 committed:
update lincense

jope35 committed:
restore .github folder

jope35 committed:
remove ruff and pre-commit from dev reqs and add ryp2

jope35 committed:
restore setup.py

jope35 committed:
restore .gitignore

jope35 committed:
create nbdev specific .gitconfig and .gitattributes

jmoralez commented:
Some more requests:
* Please update the settings.ini file with the [current settings in setup.py](https://github.com/Nixtla/tsfeatures/blob/fe4f6e63b8883f84922354b7a57056cf534aa4ae/setup.py), e.g. version should be 0.4.5. For the author and other stuff you can refer to [this file](https://github.com/Nixtla/statsforecast/blob/c929a6635a7eddd75accd860bc7765c276a8b4e8/settings.ini).
* Please update the setup.py file. I believe you can copy [this one](https://github.com/fastai/nbdev/blob/master/setup.py)

jmoralez commented on tsfeatures/__init__.py:
```diff
@@ -1,4 +1 @@
-#!/usr/bin/env python
-# coding: utf-8
-
-from .tsfeatures import *
```
Please restore this below version

jope35 replied:
added the line `from .tsfeatures import *` below the current `__version__` identifier

jmoralez commented on nbs/00_utils.ipynb:
```diff
@@ -0,0 +1,651 @@
+{
```
<p>Please place this under a #| hide, e.g.</p><pre class="ql-syntax" spellcheck="false">#| hide
%load_ext autoreload
%autoreload 2
</pre>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Nixtla/tsfeatures/pull/39/discussion/#comment-1454226253'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='2'/>

jmoralez replied:
<pre class="ql-syntax" spellcheck="false">Line #4.    from fastcore.test import *
</pre><p>This should be on a separate cell (so we don't export everything in fastcore.test from this module)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Nixtla/tsfeatures/pull/39/discussion/#comment-1454226255'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='3'/>

jmoralez replied:
<pre class="ql-syntax" spellcheck="false">Line #7.    # from scipy.signal import periodogram, welch
</pre><p>If these aren't used please just delete them</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Nixtla/tsfeatures/pull/39/discussion/#comment-1454226257'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='3'/>

jmoralez commented on nbs/01_features.ipynb:
```diff
@@ -0,0 +1,1228 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #1.    # |default_exp features
</pre><p>This should be <code>#| default_exp tsfeatures</code> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Nixtla/tsfeatures/pull/39/discussion/#comment-1454226260'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='1'/>

jope35 committed:
remove *.ipynb from gitignore

jope35 committed:
align settings and setup

jope35 committed:
refactor: Restructure notebooks and utils

Renamed and reorganized notebooks for clarity. Moved some utils to hide cells. Updated version and doc links. Added script to compare features with R. Made other minor code tweaks.

jope35 committed:
update settings with nixtla user

jmoralez commented:
Thanks for the updates. Can you please merge tsfeatures_core.ipynb with features.ipynb? They both should be the same notebook that exports the tsfeatures module.

jope35 committed:
merged core and featrues into one notebook

jope35 commented:
<p>Merge tsfeatures_core.ipynb with features.ipynb, the result is exported to tsfeatures. </p>

jmoralez commented on setup.py:
```diff
@@ -16,13 +16,13 @@
         "Operating System :: OS Independent",
     ],
-    python_requires='>=3.7',
+    python_requires=">=3.7",
```
Please replace this file with [this one](https://github.com/fastai/nbdev/blob/master/setup.py)

jope35 replied:
replaced the file

jmoralez commented on tsfeatures/tsfeatures.py:
```diff
@@ -64,17 +80,14 @@ def acf_features(x: np.array, freq: int = 1) -> Dict[str, float]:
     if size_x > 10:
         acfdiff1x = acf(np.diff(x, n=1), nlags=10, fft=False)
     else:
-        acfdiff1x = [np.nan]*2
+        acfdiff1x = [np.nan] * 2
 
     if size_x > 11:
         acfdiff2x = acf(np.diff(x, n=2), nlags=10, fft=False)
     else:
         acfdiff2x = [np.nan] * 2
     # first autocorrelation coefficient
-    try:
-        acf_1 = acfx[1]
-    except:
-        acf_1 = np.nan
+    acf_1 = acfx[1]
```
I think this was here because the result may not be of length >= 2.

jope35 replied:
gotcha

jmoralez commented on tsfeatures/tsfeatures.py:
```diff
@@ -117,12 +132,12 @@ def arch_stat(x: np.array, freq: int = 1,
         'arch_lm': R^2 value of an autoregressive model of order lags applied to x**2.
     """
     if len(x) <= lags + 1:
-        return {'arch_lm': np.nan}
+        return {"arch_lm": np.nan}
     if demean:
-        x = x - np.mean(x)
+        x -= np.mean(x)
```
This would modify the input, which is an unwanted side-effect.

jope35 replied:
reverted back

jmoralez commented on tsfeatures/tsfeatures.py:
```diff
@@ -218,12 +237,11 @@ def flat_spots(x: np.array, freq: int = 1) -> Dict[str, float]:
+        return {"flat_spots": np.nan}
 
-    return {'flat_spots': rlex}
+    rlex = np.array([sum(1 for i in g) for k, g in groupby(cutx)]).max()
```
I think there's a missing return here

jope35 replied:
add a dict with the rlex as return value

jmoralez commented on tsfeatures/tsfeatures.py:
```diff
@@ -753,43 +777,38 @@ def stl_features(x: np.array, freq: int = 1) -> Dict[str, float]:
         trough = m if trough == 0 else trough
     # Compute measure of spikiness
     d = (remainder - np.nanmean(remainder)) ** 2
-    varloo = (vare * (n-1) - d) / (n - 2)
+    varloo = (vare * (n - 1) - d) / (n - 2)
     spike = np.nanvar(varloo, ddof=1)
     # Compute measures of linearity and curvature
     time = np.arange(n) + 1
     poly_m = poly(time, 2)
     time_x = add_constant(poly_m)
     coefs = OLS(trend0, time_x).fit().params
 
-    try:
-        linearity = coefs[1]
-    except:
-        linearity = np.nan
-    try:
-        curvature = -coefs[2]
-    except:
-        curvature = np.nan
+    linearity = coefs[1]
+    curvature = -coefs[2]
```
These aren't guaranteed to have length 2 and 3 respectively

jope35 replied:
gotcha, i added the try-except stmts back in

jmoralez commented on tsfeatures/tsfeatures.py:
```diff
@@ -835,109 +855,96 @@ def unitroot_pp(x: np.array, freq: int = 1) -> Dict[str, float]:
 
-    return {'unitroot_pp': test_pp}
-
-def statistics(x: np.array, freq: int = 1) -> Dict[str, float]:
```
This function seems to be missing

jope35 replied:
added the statistics function

jmoralez replied:
You forgot the `#| export` for this function

jope35 replied:
added the export directive

jmoralez commented on tsfeatures/tsfeatures.py:
```diff
@@ -835,109 +855,96 @@ def unitroot_pp(x: np.array, freq: int = 1) -> Dict[str, float]:
+    ],
+    dict_freqs=FREQS,
+):
+    print("dict_freq")
```
Remove this print

jope35 replied:
removed the print function

jmoralez commented on tsfeatures/tsfeatures.py:
```diff
@@ -964,76 +971,14 @@ def tsfeatures(ts: pd.DataFrame,
-#### MAIN WIDE FUNCTION ########################################################
-################################################################################
-
-def _get_feats_wide(index,
-                    ts,
-                    scale = True,
-                    features = [acf_features, arch_stat, crossing_points,
-                                entropy, flat_spots, heterogeneity, holt_parameters,
-                                lumpiness, nonlinearity, pacf_features, stl_features,
-                                stability, hw_parameters, unitroot_kpss, unitroot_pp,
-                                series_length, hurst]):
-    seasonality = ts['seasonality'].item()
-    y = ts['y'].item()
-    y = np.array(y)
-
-    if scale:
-        y = scalets(y)
-
-    c_map = ChainMap(*[dict_feat for dict_feat in [func(y, seasonality) for func in features]])
-
-    return pd.DataFrame(dict(c_map), index = [index])
-
-def tsfeatures_wide(ts: pd.DataFrame,
```
These functions seem to be missing as well

jope35 replied:
include both the wide feature funcs

jmoralez commented on tsfeatures/tsfeatures.py:
```diff
@@ -1,37 +1,53 @@
+from .utils import *
 
+# %% ../nbs/tsfeatures.ipynb 7
+FREQS = {"H": 24, "D": 1, "M": 12, "Q": 4, "W": 1, "Y": 1}
```
This is defined in utils as well, please remove it from here

jope35 replied:
removed

jmoralez reviewed (changes requested):
Please also add a workflow that runs `nbdev_test`. You can refer to [this](https://github.com/Nixtla/utilsforecast/blob/ebf72a977782bb3540d826dd96d9e291415edbea/.github/workflows/ci.yaml#L43-L63)

jope35 committed:
descriptive stats on cluster results

jope35 committed:
add github action for testing

jope35 committed:
modified setup

jope35 committed:
include a pip reqs file

jope35 committed:
exclude rpy2

jope35 committed:
run nbdev_export

jmoralez commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,39 @@
+  cancel-in-progress: true
+
+jobs:
+  windows-tests:
+    runs-on: windows-latest
```
```suggestion
  run-tests:
    runs-on: [ubuntu-latest, macos-latest, windows-latest]
```

jope35 replied:
currently only test on ubuntu as the other runners are not available in a timely fashion

jmoralez replied:
What do you mean?

jope35 replied:
https://docs.github.com/en/actions/using-jobs/choosing-the-runner-for-a-job#overview

fourth bullet: "If you specify an array of strings or variables, your workflow will execute on any runner that matches all of the specified runs-on values." 
it's unlikely that a single runner exists with all the requested operating systems

I modified the workflow to create a matrix over the supported Python versions and operating systems to address this issue. This change allows your workflow to run on multiple combinations of both Python versions and OSs during each CI job execution.

jmoralez replied:
Oh yeah, sorry. That's why I meant. Thanks!

jmoralez commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,39 @@
+        run: pip install ".[dev]"
+
+      - name: Run tests
+        run: nbdev_test --do_print --timing --flags 'matplotlib polars pyarrow scipy'
```
```suggestion
        run: nbdev_test --do_print --timing
```

jmoralez commented on README.md:
```diff
@@ -1,75 +1,86 @@
-[![Build](https://github.com/FedericoGarza/tsfeatures/workflows/Python%20package/badge.svg)](https://github.com/FedericoGarza/tsfeatures/tree/master)
-[![PyPI version fury.io](https://badge.fury.io/py/tsfeatures.svg)](https://pypi.python.org/pypi/tsfeatures/)
-[![Downloads](https://pepy.tech/badge/tsfeatures)](https://pepy.tech/project/tsfeatures)
-[![Python 3.6+](https://img.shields.io/badge/python-3.7+-blue.svg)](https://www.python.org/downloads/release/python-370+/)
-[![License: MIT](https://img.shields.io/badge/License-MIT-green.svg)](https://github.com/FedericoGarza/tsfeatures/blob/master/LICENSE)
```
Please add these badges to the index notebook. [Here's](https://github.com/Nixtla/mlforecast/blob/main/nbs/index.ipynb) an example

jope35 replied:
added the badges

jmoralez commented on README.md:
```diff
@@ -118,33 +134,126 @@ python -m tsfeatures.compare_with_r --results_directory /some/path
+| diff2x_pacf5    |    0 | arch_acf          | 2.18 | entropy   | 9.45 |         |       |
 
-To replicate this results use:
+[![Build](https://github.com/FedericoGarza/tsfeatures/workflows/Python%20package/badge.svg)](https://github.com/FedericoGarza/tsfeatures/tree/master)
```
Seems like from this point onwards this is duplicated

jope35 replied:
de-duplicated

jmoralez replied:
This is still duplicated

jmoralez commented on requirements.txt:
```diff
@@ -1,6 +1 @@
-scikit-learn>=0.23.1
-statsmodels>=0.12.2
-supersmoother>=0.4
+nbdev
```
Where is this file used?

jope35 replied:
nbdev was required to install the lib, now it is placed in the gh action file

jmoralez commented on settings.ini:
```diff
@@ -0,0 +1,50 @@
+### Optional ###
+requirements = antropy>=0.1.4 arch>=4.11 pandas>=1.0.5 scikit-learn>=0.23.1 statsmodels>=0.13.2 supersmoother>=0.4 numba>=0.55.0 numpy>=1.21.6 tqdm
+dev_requirements = nbdev
+; dev_requirements = nbdev rpy2
+# console_scripts =
```
```suggestion
```

jmoralez reviewed (changes requested):
Please also move the test from the test_features notebook to the tsfeatures one

jope35 committed:
incorporate pr feedback

jope35 committed:
adding shields

jope35 committed:
only run on ubuntu

jope35 committed:
explicit ci steps to install nbdev, move test features

jope35 committed:
ci mod

jope35 committed:
no pip cache

jope35 committed:
add import stmt

jope35 committed:
add star import

jope35 committed:
small cleanup

jmoralez reviewed (changes requested):
Thanks for the changes! I think we're almost there

jope35 committed:
cleanup notebooks

jope35 committed:
add os to runs-on

jope35 committed:
change order of runs-on

jope35 committed:
matrix python version and os

jope35 committed:
optimize ci file

jope35 committed:
no cache

jmoralez commented on tsfeatures/tsfeatures_r.py:
```diff
@@ -1,155 +0,0 @@
-#!/usr/bin/env python
-# coding: utf-8
-
-from typing import List
```
Can you please add this to a new notebook nbs/tsfeatures_r.ipynb?

jope35 replied:
i added the notebook and extended the ci run to include R also added rpy2 in the dependencies list, @jmoralez maybe move it to the dev dependencies as it gives some troubles installing on windows

jmoralez commented on tsfeatures/tests/test_holt_parameters.py:
```diff
@@ -1,18 +0,0 @@
-#!/usr/bin/env python
```
Please add these tests to nbs/tsfeatures.ipynb

jope35 replied:
added the tests to the tsfeatures.ipynb

jmoralez commented on tsfeatures/tests/test_mutability.py:
```diff
@@ -1,31 +0,0 @@
-import numpy as np
```
Please add these tests to nbs/tsfeatures.ipynb

jope35 replied:
added the tests to the tsfeatures.ipynb

jmoralez commented on tsfeatures/tests/test_pacf_features.py:
```diff
@@ -1,10 +0,0 @@
-#!/usr/bin/env python
```
Please add these tests to nbs/tsfeatures.ipynb

jope35 replied:
added the tests to the tsfeatures.ipynb

jmoralez commented on tsfeatures/tests/test_small_ts.py:
```diff
@@ -1,36 +0,0 @@
-import numpy as np
```
Please add these tests to nbs/tsfeatures.ipynb

jope35 replied:
added the tests to the tsfeatures.ipynb

jmoralez commented on tsfeatures/tests/test_sparsity.py:
```diff
@@ -1,25 +0,0 @@
-#!/usr/bin/env python
```
Please add these tests to nbs/tsfeatures.ipynb

jope35 replied:
added the tests to the tsfeatures.ipynb

jmoralez commented on tsfeatures/tests/test_statistics.py:
```diff
@@ -1,25 +0,0 @@
-#!/usr/bin/env python
```
Please add these tests to nbs/tsfeatures.ipynb

jope35 replied:
added the tests to the tsfeatures.ipynb

jope35 committed:
include tests and R CI

jope35 committed:
update readme

jope35 committed:
alter rpy2 version

jope35 committed:
debug R install

jope35 committed:
more debug

jope35 committed:
more debug2

jope35 committed:
one string

jope35 committed:
asdf

jope35 committed:
alter rpy2 version

jope35 committed:
update R version

jope35 committed:
exclude the R dependency

jope35 committed:
other R settings

jope35 committed:
explicit PATH

jope35 committed:
echo rhome

jope35 committed:
export R functions

jope35 committed:
include ubuntu

jope35 committed:
include fastcore explicitly as a dev dep

jope35 committed:
no parallel

jope35 committed:
Merge branch 'main' into main
